patch/gopatch.go

```diff
@@ -0,0 +1,133 @@
+package patch
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/parser"
+	"go/token"
+	"golang.org/x/tools/imports"
+	"sort"
+
+	"github.com/uber-go/gopatch/internal/astdiff"
+	"github.com/uber-go/gopatch/internal/engine"
+	"github.com/uber-go/gopatch/internal/parse"
+)
+
+// PatchFile is a patch difference that can be applied to Go file.
+type PatchFile struct {
+	fset *token.FileSet
+	prog *engine.Program
+}
+
+// Parse the patch file and creates data that can be applied to the Go file.
+func Parse(patchFileName string, src []byte) (*PatchFile, error) {
+	fset := token.NewFileSet()
+
+	astProg, err := parse.Parse(fset, patchFileName, src)
+	if err != nil {
+		return nil, fmt.Errorf("parse: %w", err)
+	}
+	prog, err := engine.Compile(fset, astProg)
+	if err != nil {
+		return nil, fmt.Errorf("compile: %w", err)
+	}
+
+	return &PatchFile{fset: fset, prog: prog}, nil
+}
+
+// Apply takes the Go file name and its contents and returns a Go file with the patch applied.
+func (patchFile *PatchFile) Apply(filename string, src []byte) ([]byte, error) {
```

Comment:
small nit: please use a short receiver name here.

```suggestion
func (pf *PatchFile) Apply(filename string, src []byte) ([]byte, error) {
```

Related: https://go.dev/wiki/CodeReviewComments#receiver-names

Reply:
did it 👍